python/ray/_private/utils.py

```diff
@@ -1775,3 +1777,77 @@ def _get_pyarrow_version() -> Optional[str]:
             if hasattr(pyarrow, "__version__"):
                 _PYARROW_VERSION = pyarrow.__version__
     return _PYARROW_VERSION
+
+
+class DeferSigint(contextlib.AbstractContextManager):
+    """Context manager that defers SIGINT signals until the the context is left."""
+
+    # This is used by Ray's task cancellation to defer cancellation interrupts during
+    # problematic areas, e.g. task argument deserialization.
+    def __init__(self):
+        # Whether the task has been cancelled while in the context.
+        self.task_cancelled = False
+        # The original SIGINT handler.
+        self.orig_sigint_handler = None
+        # The original signal method.
+        self.orig_signal = None
+
+    @classmethod
+    def create_if_main_thread(cls) -> contextlib.AbstractContextManager:
+        """Creates a DeferSigint context manager if running on the main thread,
+        returns a no-op context manager otherwise.
+        """
+        if threading.current_thread() == threading.main_thread():
+            return cls()
+        else:
+            # TODO(Clark): Use contextlib.nullcontext() once Python 3.6 support is
+            # dropped.
+            return contextlib.suppress()
+
+    def _set_task_cancelled(self, signum, frame):
+        """SIGINT handler that defers the signal."""
+        self.task_cancelled = True
+
+    def _signal_monkey_patch(self, signum, handler):
+        """Monkey patch for signal.signal that raises an error if a SIGINT handler is
+        registered within the DeferSigint context.
+        """
+        # Only raise an error if setting a SIGINT handler in the main thread; if setting
+        # a handler in a non-main thread, signal.signal will raise an error anyway
+        # indicating that Python does not allow that.
+        if (
+            threading.current_thread() == threading.main_thread()
+            and signum == signal.SIGINT
+        ):
+            raise ValueError(
+                "Can't set signal handler for SIGINT while SIGINT is being deferred "
+                "within a DeferSigint context."
+            )
+        return self.orig_signal(signum, handler)
+
+    def __enter__(self):
+        # Save original SIGINT handler for later restoration.
+        self.orig_sigint_handler = signal.getsignal(signal.SIGINT)
+        # Set SIGINT signal handler that defers the signal.
+        signal.signal(signal.SIGINT, self._set_task_cancelled)
+        # Monkey patch signal.signal to raise an error if a SIGINT handler is registered
+        # within the context.
+        self.orig_signal = signal.signal
+        signal.signal = self._signal_monkey_patch
```

Comment:
how reliable is this monkey patch? 
think about it again, an alternative is to defer sending signals when serialization is in progress: 
https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/_raylet.pyx?L1210

Reply:
This isn't technically needed for our current use of the context manager on task argument deserialization, this is just to:
1. return a nice error to users that are doing something really weird, like registering a signal handler in the deserialization-side of a user-defined type pickle reducer,
2. guard against future uses of this context manager elsewhere in Ray Core.

Reply:
>think about it again, an alternative is to defer sending signals when serialization is in progress:
https://sourcegraph.com/github.com/ray-project/ray/-/blob/python/ray/_raylet.pyx?L1210

Yeah I thought of that, but I thought that deferring the signal to the end of the code block in question would be cleaner than adding another mutex-guarded `is_deserializing_args` flag and relying on the task cancellation RPC retry.

In my testing, I think that I hit some hangs and silent failures when task cancellation was received before the task ID was set (i.e. with the RPC retry being hit), so I'm worried that there are some undiscovered bugs there. Going to dig a bit further into that to see if I can get a reproduction.